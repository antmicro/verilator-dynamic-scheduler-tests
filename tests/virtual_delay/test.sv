virtual class Delay;
    pure virtual task do_delay;
endclass

`define DELAY(dt) \
class Delay``dt extends Delay; \
    virtual task do_delay; \
        $write("Starting a #%0d delay\n", dt); \
        #dt \
        $write("Ended a #%0d delay\n", dt); \
    endtask \
endclass

`DELAY(10);
`DELAY(20);
`DELAY(40);
`DELAY(80);

module t;
    task printtime;
        $write("I'm at time %0t\n", $time);
    endtask

    initial begin
        Delay delay;
        Delay10 d10 = new;
        Delay20 d20 = new;
        Delay40 d40 = new;
        Delay80 d80 = new;
        printtime;
        delay = d10;
        delay.do_delay;
        printtime;
        delay = d20;
        delay.do_delay;
        printtime;
        delay = d40;
        delay.do_delay;
        printtime;
        delay = d80;
        delay.do_delay;
        printtime;
        $write("*-* All Finished *-*\n");
        $finish;
    end
endmodule
