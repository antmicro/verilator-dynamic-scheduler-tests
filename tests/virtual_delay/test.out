I'm at time 0
Starting a #10 delay
I'm at time 10
Starting a #20 delay
I'm at time 30
Starting a #40 delay
I'm at time 70
Starting a #80 delay
I'm at time 150
*-* All Finished *-*
